lib/promise/group.rb

```diff
@@ -0,0 +1,39 @@
+class Promise
+  class Group
+    attr_reader :promise
+
+    def initialize(inputs)
+      @remaining = 0
+      @promise = Promise.new
+      @inputs = inputs
+      @on_fulfill = method(:on_fulfill)
+      @on_reject = promise.method(:reject)
+      chain_inputs(inputs)
+    end
+
+    private
+
+    def chain_inputs(inputs)
+      inputs.each do |obj|
+        if promise?(obj)
+          @remaining += 1
+          obj.then(@on_fulfill, @on_reject)
```

Comment:
We can't `then` this right away because if `obj` is already resolved it may call `on_fulfill` inline, at which point `@remaining` could return to 0 even if there are unfulfilled promises later on in the inputs array. We should `select` the inputs that are actually promises, then add the count to `@remaining` all at once before setting up our `then`s.